documentation/prometheus-design-principles.md

```diff
@@ -0,0 +1,164 @@
+
+## Shape of data
+
+### Names
+
+  * tags should not have too high a cardinality
+      * See the CAUTION here: https://prometheus.io/docs/practices/naming/#labels
+      * it seems that prometheus 2.0's new storage engine took a lot
+      of effort to support "churn" of tag labels, so long as the
+      number of active labels at any point is relatively small. see
+      ["storing 16 bytes at scale" from promcon 2017](https://promcon.io/2017-munich/talks/storing-16-bytes-at-scale/)
+      or
+      [prometheus 2.0: new storage layer...](https://coreos.com/blog/prometheus-2.0-storage-layer-optimization)
+  * tags should not include redundant informational data
+
+### Values
+
+  * [counters should always count up from zero](https://www.robustperception.io/how-does-a-prometheus-counter-work/)
+      * data should always make sense, independent of scrape interval
+      * this is in contrast to statsd's behaviour
+  * [sum of rates, not rate of sums](https://www.robustperception.io/rate-then-sum-never-sum-then-rate/)
+
+## Architectural things ##
+
+### Simplicity ###
+
+Prometheus is designed to achieve reliability through simplicity.
+
+This means:
+  * redundancy is achieved by deploying multiple separate instances,
+    not by database replication.
+  * recovery from backup is not a priority.  if a prometheus server
+    has a disk failure, just use another redundant instance (assuming
+    you have one)
+  * if a prometheus instance has a temporary blip (say, a network
+    issue) and then recovers, it will have missing metrics for the
+    duration of the issue.
+    
+[Julius Volz talked about why they didn't do clustering](https://youtu.be/QgJbxCWRZ1s?t=26m40s).
+
+### Prometheus should be deployed in same failure domain ###
+
+A prometheus server should be deployed in the same failure domain as
+the things it monitors.  This means the same data centre, region,
+availability zone, etc.
+
+This means that:
+
+  * a prometheus shouldn't have to go over the public internet to
+    scrape a service
+      * (federation might go over public internet, though)
+      
+https://www.robustperception.io/scaling-and-federating-prometheus/
+
+### There should normally be a single alertmanager for an organization ###
+
+It seems that there should normally be a single alertmanager for an
+organization.  The best reference I've found for this is from
+[Julius Volz's Prometheus Design and Philosophy talk (7:44)](https://youtu.be/QgJbxCWRZ1s?t=7m43s)
+which doesn't go into much detail as to *why* you'd do this.
+
+## Security model
+
+Prometheus has extensive
+[documentation about its security model](https://prometheus.io/docs/operating/security/).
+In particular, it distinguishes trusted users (who have admin access
+to deploy and reconfigure prometheus) from untrusted users (who can
+view prometheus metrics and dashboards).
+
+Mostly, you control access to prometheus at the network level; there
```

Comment:
We are also able to leave the HTTP write-API disabled (as it is by default)

Reply:
@davbo do you mean the `--web.enable-admin-api` option? or something else?

Reply:
yep that's the one

Reply:
great, that means I got it covered by 326fd17